core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala

```diff
@@ -0,0 +1,237 @@
+package com.mesosphere.usi.core
+
+import java.time.Instant
+
+import com.mesosphere.usi.core.models._
+import com.mesosphere.usi.models._
+
+case class FrameWithEffects(frame: Frame, effects: FrameEffects, dirtyPodIds: Set[PodId]) {
+  def applyEffects(newEffects: FrameEffects): FrameWithEffects = {
+    // TODO - we need to handle status snapshots and create a mechanism to signal that all cache should be recomputed
+    val newDirty = dirtyPodIds ++ newEffects.reverseStateEvents.iterator.collect {
+      case podEvent: PodStateEvent => podEvent.id
+    }
+    copy(frame = frame.applyStateEffects(effects), dirtyPodIds = newDirty, effects = effects ++ newEffects)
```

Comment:
Shouldn't the original effects be dropped since they are already applied?

Reply:
Good question. However, of these effects, only `stateEvents` (and not e.g. `mesosCalls`) are applied. So I guess we should keep the later and remove the former?

Reply:
We have to aggregate the events.